MAINT: `matplotlib==3.5b1` causes DXT heatmap code failures

The failures are caused by `python setup.py install` pulling in a bleeding-edge version of `matplotlib` (`3.5b1`). Longer-term we should probably modernize the build system per gh-476, but that's a separate matter (to `pip install .` with the `autoperf` stuff enabled, and how much "Python stuff" should still be done by the automake machinery).

I'll place the failures below the fold. Questions to answer might be:

1) Is this caused by `seaborn` not yet having shims for this bleeding-edge version of `matplotlib`?
2) Is this something we need to/should handle ourselves because `seaborn` gives us different handles with different versions of `matplotlib`?
(that kind of thing..)

For now, I will pin the `matplotlib` version we use in CI, since it is still effectively labelled as a beta for example.

Here are the tracebacks in case it helps:

<details>

```
========================================================================================================================================================= FAILURES =========================================================================================================================================================
_____________________________________________________________________________________________________________________________ test_adjust_for_colorbar[tests/input/sample-dxt-simple.darshan] ______________________________________________________________________________________________________________________________
[gw6] darwin -- Python 3.8.1 /Users/treddy/python_venvs/python_381_venv/bin/python

filepath = 'tests/input/sample-dxt-simple.darshan'

    @pytest.mark.parametrize(
        "filepath",
        [
            "examples/example-logs/ior_hdf5_example.darshan",
            "examples/example-logs/dxt.darshan",
            "tests/input/sample-dxt-simple.darshan",
        ],
    )
    def test_adjust_for_colorbar(filepath):
        # regression test for `plot_dxt_heatmap.adjust_for_colorbar()`
    
        jgrid = plot_dxt_heatmap.plot_heatmap(log_path=filepath)
    
        # the plot positions change based on the number of unique ranks.
        # If there is only 1 rank, there is no horizontal bar graph
        # so the x-axis values are scaled accordingly.
    
        # get heatmap positions
        hmap_positions = jgrid.ax_joint.get_position()
        assert hmap_positions.x0 == 0.1
        assert hmap_positions.y0 == 0.15000000000000002
        assert hmap_positions.y1 == 0.774390243902439
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, it has
            # different x max values because it needs room for
            # the colorbar on the outside of the horizontal bar graph
            assert hmap_positions.x1 == 0.7158709677419354
        else:
            assert hmap_positions.x1 == 0.7824516129032258
    
        # get vertical bar graph positions
        vert_bar_positions = jgrid.ax_marg_x.get_position()
        assert vert_bar_positions.x0 == 0.1
        assert vert_bar_positions.y0 == 0.7780487804878049
        assert vert_bar_positions.y1 == 0.9
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, the vertical
            # bar graph has a different x max value because it needs room for
            # the colorbar on the outside of the horizontal bar graph
            assert vert_bar_positions.x1 == 0.7158709677419354
        else:
            assert vert_bar_positions.x1 == 0.7824516129032258
    
        # get horizontal bar graph positions
        horiz_bar_positions = jgrid.ax_marg_y.get_position()
        assert horiz_bar_positions.y0 == 0.15000000000000002
        assert horiz_bar_positions.y1 == 0.774390243902439
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, the horizontal
            # bar graph has different x min/max values because it has to
            # make room for the colorbar
            assert horiz_bar_positions.x0 == 0.7206451612903225
            assert horiz_bar_positions.x1 == 0.84
        else:
            assert horiz_bar_positions.x0 == 0.7877419354838711
            assert horiz_bar_positions.x1 == 0.92
    
        # get the colorbar positions
        cbar_positions = jgrid.fig.axes[-1].get_position()
        assert cbar_positions.y0 == 0.15000000000000002
        assert cbar_positions.y1 == 0.774390243902439
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, the colorbar has
            # to go closer to the edge of the figure so the horizontal bar
            # graph can fit in the panel
            assert cbar_positions.x0 == 0.85
            assert cbar_positions.x1 == 0.8716135084427767
        else:
            assert cbar_positions.x0 == 0.82
>           assert cbar_positions.x1 == 0.8416135084427767
E           assert 1.72 == 0.8416135084427767
E            +  where 1.72 = Bbox([[0.82, 0.15000000000000002], [1.72, 0.774390243902439]]).x1

tests/test_plot_dxt_heatmap.py:397: AssertionError
_________________________________________________________________________________________________________________________ test_adjust_for_colorbar[examples/example-logs/ior_hdf5_example.darshan] _________________________________________________________________________________________________________________________
[gw0] darwin -- Python 3.8.1 /Users/treddy/python_venvs/python_381_venv/bin/python

filepath = 'examples/example-logs/ior_hdf5_example.darshan'

    @pytest.mark.parametrize(
        "filepath",
        [
            "examples/example-logs/ior_hdf5_example.darshan",
            "examples/example-logs/dxt.darshan",
            "tests/input/sample-dxt-simple.darshan",
        ],
    )
    def test_adjust_for_colorbar(filepath):
        # regression test for `plot_dxt_heatmap.adjust_for_colorbar()`
    
        jgrid = plot_dxt_heatmap.plot_heatmap(log_path=filepath)
    
        # the plot positions change based on the number of unique ranks.
        # If there is only 1 rank, there is no horizontal bar graph
        # so the x-axis values are scaled accordingly.
    
        # get heatmap positions
        hmap_positions = jgrid.ax_joint.get_position()
        assert hmap_positions.x0 == 0.1
        assert hmap_positions.y0 == 0.15000000000000002
        assert hmap_positions.y1 == 0.774390243902439
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, it has
            # different x max values because it needs room for
            # the colorbar on the outside of the horizontal bar graph
            assert hmap_positions.x1 == 0.7158709677419354
        else:
            assert hmap_positions.x1 == 0.7824516129032258
    
        # get vertical bar graph positions
        vert_bar_positions = jgrid.ax_marg_x.get_position()
        assert vert_bar_positions.x0 == 0.1
        assert vert_bar_positions.y0 == 0.7780487804878049
        assert vert_bar_positions.y1 == 0.9
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, the vertical
            # bar graph has a different x max value because it needs room for
            # the colorbar on the outside of the horizontal bar graph
            assert vert_bar_positions.x1 == 0.7158709677419354
        else:
            assert vert_bar_positions.x1 == 0.7824516129032258
    
        # get horizontal bar graph positions
        horiz_bar_positions = jgrid.ax_marg_y.get_position()
        assert horiz_bar_positions.y0 == 0.15000000000000002
        assert horiz_bar_positions.y1 == 0.774390243902439
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, the horizontal
            # bar graph has different x min/max values because it has to
            # make room for the colorbar
            assert horiz_bar_positions.x0 == 0.7206451612903225
            assert horiz_bar_positions.x1 == 0.84
        else:
            assert horiz_bar_positions.x0 == 0.7877419354838711
            assert horiz_bar_positions.x1 == 0.92
    
        # get the colorbar positions
        cbar_positions = jgrid.fig.axes[-1].get_position()
        assert cbar_positions.y0 == 0.15000000000000002
        assert cbar_positions.y1 == 0.774390243902439
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, the colorbar has
            # to go closer to the edge of the figure so the horizontal bar
            # graph can fit in the panel
            assert cbar_positions.x0 == 0.85
>           assert cbar_positions.x1 == 0.8716135084427767
E           assert 1.75 == 0.8716135084427767
E            +  where 1.75 = Bbox([[0.85, 0.15000000000000002], [1.75, 0.774390243902439]]).x1

tests/test_plot_dxt_heatmap.py:394: AssertionError
_______________________________________________________________________________________________________________________________ test_adjust_for_colorbar[examples/example-logs/dxt.darshan] ________________________________________________________________________________________________________________________________
[gw4] darwin -- Python 3.8.1 /Users/treddy/python_venvs/python_381_venv/bin/python

filepath = 'examples/example-logs/dxt.darshan'

    @pytest.mark.parametrize(
        "filepath",
        [
            "examples/example-logs/ior_hdf5_example.darshan",
            "examples/example-logs/dxt.darshan",
            "tests/input/sample-dxt-simple.darshan",
        ],
    )
    def test_adjust_for_colorbar(filepath):
        # regression test for `plot_dxt_heatmap.adjust_for_colorbar()`
    
        jgrid = plot_dxt_heatmap.plot_heatmap(log_path=filepath)
    
        # the plot positions change based on the number of unique ranks.
        # If there is only 1 rank, there is no horizontal bar graph
        # so the x-axis values are scaled accordingly.
    
        # get heatmap positions
        hmap_positions = jgrid.ax_joint.get_position()
        assert hmap_positions.x0 == 0.1
        assert hmap_positions.y0 == 0.15000000000000002
        assert hmap_positions.y1 == 0.774390243902439
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, it has
            # different x max values because it needs room for
            # the colorbar on the outside of the horizontal bar graph
            assert hmap_positions.x1 == 0.7158709677419354
        else:
            assert hmap_positions.x1 == 0.7824516129032258
    
        # get vertical bar graph positions
        vert_bar_positions = jgrid.ax_marg_x.get_position()
        assert vert_bar_positions.x0 == 0.1
        assert vert_bar_positions.y0 == 0.7780487804878049
        assert vert_bar_positions.y1 == 0.9
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, the vertical
            # bar graph has a different x max value because it needs room for
            # the colorbar on the outside of the horizontal bar graph
            assert vert_bar_positions.x1 == 0.7158709677419354
        else:
            assert vert_bar_positions.x1 == 0.7824516129032258
    
        # get horizontal bar graph positions
        horiz_bar_positions = jgrid.ax_marg_y.get_position()
        assert horiz_bar_positions.y0 == 0.15000000000000002
        assert horiz_bar_positions.y1 == 0.774390243902439
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, the horizontal
            # bar graph has different x min/max values because it has to
            # make room for the colorbar
            assert horiz_bar_positions.x0 == 0.7206451612903225
            assert horiz_bar_positions.x1 == 0.84
        else:
            assert horiz_bar_positions.x0 == 0.7877419354838711
            assert horiz_bar_positions.x1 == 0.92
    
        # get the colorbar positions
        cbar_positions = jgrid.fig.axes[-1].get_position()
        assert cbar_positions.y0 == 0.15000000000000002
        assert cbar_positions.y1 == 0.774390243902439
        if filepath == "examples/example-logs/ior_hdf5_example.darshan":
            # since `ior_hdf5_example.darshan` has 4 ranks, the colorbar has
            # to go closer to the edge of the figure so the horizontal bar
            # graph can fit in the panel
            assert cbar_positions.x0 == 0.85
            assert cbar_positions.x1 == 0.8716135084427767
        else:
            assert cbar_positions.x0 == 0.82
>           assert cbar_positions.x1 == 0.8416135084427767
E           assert 1.72 == 0.8416135084427767
E            +  where 1.72 = Bbox([[0.82, 0.15000000000000002], [1.72, 0.774390243902439]]).x1

tests/test_plot_dxt_heatmap.py:397: AssertionError
```
</details>

--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -8,8 +8,7 @@ task:
     /opt/homebrew/opt/python@3.10/bin/python3 -m venv ~/py_310
     source ~/py_310/bin/activate
     python -m pip install --upgrade pip
-    python -m pip install --upgrade pytest lxml
-    python -m pip install 'matplotlib<3.5'
+    python -m pip install --upgrade pytest lxml matplotlib packaging
     brew install automake libtool binutils
     mkdir -p /tmp/darshan_install
     export DARSHAN_INSTALL_PATH=/tmp/darshan_install

--- a/.github/workflows/main_ci.yml
+++ b/.github/workflows/main_ci.yml
@@ -31,10 +31,7 @@ jobs:
       - name: Install dependencies
         run: |
           python -m pip install --upgrade pip
-          python -m pip install --upgrade pytest mypy pyflakes asv pytest-cov codecov lxml
-          # matplotlib is pinned because of
-          # gh-479
-          python -m pip install 'matplotlib<3.5'
+          python -m pip install --upgrade pytest mypy pyflakes asv pytest-cov codecov lxml matplotlib packaging
       - if: ${{matrix.platform == 'macos-latest'}}
         name: Install MacOS deps
         run: |

--- a/darshan-util/pydarshan/darshan/tests/test_plot_dxt_heatmap.py
+++ b/darshan-util/pydarshan/darshan/tests/test_plot_dxt_heatmap.py
@@ -3,7 +3,9 @@ import numpy as np
 from numpy.testing import assert_array_equal, assert_allclose
 import pandas as pd
 import seaborn as sns
+import matplotlib
 import matplotlib.pyplot as plt
+from packaging import version
 
 import darshan
 from darshan.experimental.plots import heatmap_handling, plot_dxt_heatmap
@@ -389,10 +391,16 @@ def test_adjust_for_colorbar(filepath):
         # since `dxt.darshan` has 1 rank, the colorbar doesn't have
         # to go closer to the edge of the figure
         assert cbar_positions.x0 == 0.82
-        assert cbar_positions.x1 == 0.8416135084427767
+        if version.parse(matplotlib.__version__) < version.parse("3.5.0"):
+            assert cbar_positions.x1 == 0.8416135084427767
+        else:
+            assert cbar_positions.x1 == 1.72
     else:
         assert cbar_positions.x0 == 0.85
-        assert cbar_positions.x1 == 0.8716135084427767
+        if version.parse(matplotlib.__version__) < version.parse("3.5.0"):
+            assert cbar_positions.x1 == 0.8716135084427767
+        else:
+            assert cbar_positions.x1 == 1.75
 
 @pytest.mark.parametrize(
     "filepath",

--- a/darshan-util/pydarshan/pyproject.toml
+++ b/darshan-util/pydarshan/pyproject.toml
@@ -14,9 +14,10 @@ skip = [
     "*_s390x"
 ]
 test-requires = [
+    "packaging",
     "pytest",
     "lxml",
-    "matplotlib<3.5",
+    "matplotlib",
     "importlib_resources;python_version<'3.9'"
 ]
 before-test = "pip install -U git+https://github.com/darshan-hpc/darshan-logs.git@main"

